core/gate_hooks/track.cc

```diff
@@ -44,33 +44,32 @@ const GateHookCommands Track::cmds = {{"reset", "EmptyArg",
 Track::Track()
     : bess::GateHook(Track::kName, Track::kPriority),
       track_bytes_(),
-      cnt_(),
-      pkts_(),
-      bytes_() {}
+      worker_stats_() {}
 
 CommandResponse Track::Init(const bess::Gate *, const bess::pb::TrackArg &arg) {
   track_bytes_ = arg.bits();
   return CommandSuccess();
 }
 
 CommandResponse Track::CommandReset(const bess::pb::EmptyArg &) {
-  cnt_ = 0;
-  pkts_ = 0;
-  bytes_ = 0;
+  worker_stats_ = {};
   return CommandSuccess();
 }
 
 void Track::ProcessBatch(const bess::PacketBatch *batch) {
+
+  TrackStats *stat = &worker_stats_[current_worker.wid()];
+
   size_t cnt = batch->cnt();
-  cnt_ += 1;
-  pkts_ += cnt;
+  stat->cnt += 1;
+  stat->pkts += cnt;
 
   if (!track_bytes_) {
     return;
   }
 
   for (size_t i = 0; i < cnt; i++) {
-    bytes_ += batch->pkts()[i]->data_len() + kEthernetOverhead;
```

Comment:
My hunch is that here `stat->bytes` may be loaded and stored every iteration in the loop, instead of doing accumulation with a register. Could you use a local variable and make a quick performance comparison, with `perftest/chain.bess` and byte counting enabled?

Reply:
```
for (size_t i=0; i < cnt; i++)
   stat->bytes += batch->pkts()[i]->data_len() + kEthernetOverhead;
}
```
Chain/ 1 pps:     26390664	counts/s:       824708
Chain/ 2 pps:     13860911	counts/s:       433153
Chain/ 4 pps:      7153635	counts/s:       223551
Chain/ 8 pps:      3719552	counts/s:       116236
Chain/16 pps:      1879005	counts/s:        58719
Chain/32 pps:       941429	counts/s:        29420

```
uint64_t bytes = 0;
for (size_t i = 0; i < cnt; i++) {
  bytes += batch->pkts()[i]->data_len() + kEthernetOverhead;
}
stat->bytes += bytes;
```

Chain/ 1 pps:     26657973	counts/s:       833062
Chain/ 2 pps:     14204597	counts/s:       443894
Chain/ 4 pps:      7272441	counts/s:       227264
Chain/ 8 pps:      3719219	counts/s:       116226
Chain/16 pps:      1878811	counts/s:        58713
Chain/32 pps:       942076	counts/s:        29440

Guess Compilers do the rock here!